### ISSUE TITLE
Display full LN name

### ISSUE THREAD

**Issue by Sander3003** [openscd/open-scd#365]
As an system operator
I want to see the full LN instance name
So I don't need to remember the LN abbreviation

Background;
IEC61850 configurations could be hard to understand for non specialists. Writing the full name instead of the abbreviation will help to understand the IEC61850 configurations more easily.
The Logical node names are described in the IEC61850-7-4.

**Comment by Sander3003** [comment]
@JakobVogelsang is their a list of LN abbreviations and their full name we can use here?

**Comment by JakobVogelsang** [comment]
Hi Sander,

We don't have names for it. What we have is the `titleId´ in the NSD files. I have never had a serious look. But my guess  is that there are special files for that as well. But in any case there are only 92 LNs defined in the 7-4. We can take over the names from the PDF. This is a good idea!

**Comment by JakobVogelsang** [comment]
@Sander3003 I have defined an issue that I think will cover this as well. #389. If you are of the same opinion, please close this one. If not, we can extend #389 to cover this.

**Comment by Sander3003** [comment]
@Randy993 can you look at this?

**Comment by randy993** [comment]
We are curently looking into this. The file required is IEC_61850-7-4_2007B-en.nsdoc, however this isn't publicly available (only for purchasers of IEC 61850 7-4). We will have a look into the legal restrictions and possibilities regarding this file, but that will probably take a while. Therefore, we should consider another way of implementing this, for example an upload function per user or reading a file location which is configurable.

**Comment by danyill** [comment]
As far as I'm aware this is freely available under the TC57 support documents as `IEC_61850-7-4.NSD.2007B3.light.zip` and inside this zip file: `IEC_61850-7-4_2007B3.nsd`.

See:  https://www.iec.ch/dyn/www/f?p=103:227:0::::FSP_ORG_ID,FSP_LANG_ID:1273,25

**Comment by danyill** [comment]
I've looked into this a little more. The difference between `light` and `full` appears to be the full description of each term which is presumably exactly what we are looking for.

`IEC_61850-7-4_2007B-en.nsdoc` 

IANAL but the file above provides copyright and licensing information:

```xml
<Copyright>
         <Notice>
			COPYRIGHT (c) IEC, www.iec.ch/tc57/supportdocuments. This version of this NSD is part of IEC_61850-7-4:2007; see the IEC_61850-7-4:2007 for full legal notices. In case of any differences between the here-below code and the IEC published content, the here-below definition supersedes the IEC publication; it may contain updates. See history files. The whole document has to be taken into account to have a full description of this code component.
			See www.iec.ch/CCv1 for copyright details. 
        </Notice>
    <License uri="www.iec.ch/CCv1" kind="Standard"/>
</Copyright>
```

www.iec.ch/CCv1 redirects to https://www.iec.ch/webstore/custserv/pdf/CC-EULA.pdf which states amongst other things:

> The user of Code Components shall attribute each such Code Component to IEC and identify
> the IEC standard from which it is taken. Such attribution (e.g., “This code was derived from IEC
> [insert standard reference number:publication year] within modifications permitted in the
> relevant IEC standard. Please reproduce this note if possible.”), may be placed in the code itself
> or any other reasonable location.

So I think we need to attribute this to the appropriate standard (for instance with an entry in the help section and the README for good measure) but otherwise, it's at least arguable we are free to use and incorporate.

The IEC TC57 guidance for management of code components ([here](https://www.iec.ch/public/tc57/Handling%20of%20Code%20Components%20v8.0%202019-08-29.pdf)) describes the intention between `light` and `full` is to ensure you have to have the standard to really understand what is going on:

> 6.3.2 Full MIB and light MIB
> Two set of MIBs may be provided:
> • A set of MIBs that includes in the DESCRIPTION field of each object the copy of the description provided in the standard. This set of MIBs is more suitable for the inclusion on DRAFT NSM central tools because it provides the direct description of each object semantics to the operator.
> • A set of “light” MIBs that provides a reference to the UML object in the object description without including the object description. This approach allows public access to the MIBs while still ensuring that the IEC standard is needed to understand what the MIBs mean.
> The MIB files are included in a package together with a general license (EULA). This package contains the following files:
> • The standard document in PDF format
> • A directory containing MIBs
> • A directory containing light MIBs
> • The End User License Agreement (EULA) text file.

However Annex A of the treatment of Code Components also indicates that the intention is _only_ be that someone involved in the development of the software has to own a copy of the standard to legally include the file in the software

> “The purchase of this IEC standard carries a copyright license for the purchaser to sell software containing Code Components from this standard directly to end users and to end users via distributors, subject to IEC software licensing conditions, which can be found here: www.iec.ch/CCv1“

As far as I can tell this is compatible with an Apache 2 license.

At face value, as long as someone involved in this software owned a copy of the standard we could include the file on this basis. I understand the express intent of the `nsd` files is to facilitate the development of tools, as discussed in IEC 61850-7-7 Ed 1.0 in the Introduction:

> In order to foster an active tool market with good quality, and at the end to improve IEC 61850
> interoperability, a machine-processable file, describing data model related parts of the
> standard as input, is needed. This is the purpose of the new language Name Space Definition
> (NSD) defined by this part of IEC 61850.
> 
> This will avoid the need for any engineering tool related to the IEC 61850 data models to get
> the content of the standard manually entered, with a high risk of mistakes. This will also help
> to easily spread any corrections to the data model, as requested to reach interoperability.
> Tool vendors will be able to integrate NSD in their tools to distribute the standard data models
> directly to end users.

We already have the CC-EULA file in this repo (https://github.com/openscd/open-scd/blob/main/CC-EULA.pdf) for the XSD files. 
The light NSD is already included in this repo: https://github.com/openscd/open-scd/blob/main/public/xml/IEC_61850-7-4_2007B3.nsd and the `xsd` files are included in https://github.com/openscd/open-scd/blob/13fbd1873f486aabc1e7c8de58b7544cf7da2eb8/src/schemas.ts. However, all of these are "freely" available in the TC57 support area.

It does all appear unnecessarily complex, so perhaps a legal opinion is required but in the meantime, an extension that enriched data (perhaps with a tooltip) based on an uploaded file is probably a good idea.

**Comment by Sander3003** [comment]
Thanks @danyill for looking into this topic. My collegue @nicorikken is looking into this topic as well.

**Comment by nicorikken** [comment]
@danyill I thanks for writing down the details, I came to a similar analysis. We're now reaching out to the standards body (in our case NEN) to discuss details and explore the options.

**Comment by randy993** [comment]
As previously mentioned, in parallel to reaching out to the NEN and/or IEC, we came up with a (temporary) alternative to support showing the functional names in CoMPAS. We've made a proposal for an upload functionality of the NSDOC file for the users, when it is available. @JakobVogelsang will write the two technical stories required for the upload functionality. This story can be used to log the changes related to the IED editor.

- Class name should be changed to the friendly name when the NSDOC file is available
-- {classname}.__cl.title should be used
- DO name should be changed to the friendly name when the NSDOC file is available
-- AbbrTermB.{doname}.desc should be used
- We need a toggle button (in the settings/IED editor?) for toggling between the class/do names and the friendly names

There are still a few open questions:
- In some cases, the DO appears to be a concatination of two abbreviations. For example; MltLev. In the NSDOC file, there is no description for MltLev, but there is a seperate description for Mlt and for Lev. Maybe we should consider splitting the strings based on capitalization?
- There are descriptions available for DOs within a certain class. For example: LLN0.Beh.dsc = "Reflects the state of the logical device related hardware and software. It is derived from the worst (most critical) value of the health attribute of all the domain logical nodes contained in the logical device: 'Health' = max('DomainLN[i].Health')". I think we should show these descriptions to the user either on mouseover of the DO or as a text element when the object is expanded.
![image](https://user-images.githubusercontent.com/47896624/149734448-5c14cf11-360b-466a-82f9-2adc39328967.png)
